File: adapters/database/src/main/java/net/onelitefeather/pandorascluster/database/models/chunk/ClaimedChunkEntity.java
package net.onelitefeather.pandorascluster.database.models.chunk;

import jakarta.persistence.*;
import net.onelitefeather.pandorascluster.database.models.land.LandAreaEntity;

import java.util.Objects;

@Entity
@Table(
        name = "land_chunks",
        uniqueConstraints = @UniqueConstraint(
                name = "uk_land_chunks_area_chunk",
                columnNames = {"landArea_id", "chunk_index"}
        )
)
public final class ClaimedChunkEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "chunk_index", nullable = false)
    private Long chunkIndex;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "landArea_id", nullable = false)
    private LandAreaEntity landArea;

    public ClaimedChunkEntity() {
        // Empty constructor for Hibernate
    }

    public ClaimedChunkEntity(Long id, Long chunkIndex, LandAreaEntity landArea) {
        this.id = id;
        this.chunkIndex = chunkIndex;
        this.landArea = landArea;
    }

    public Long id() {
        return id;
    }

    public Long chunkIndex() {
        return chunkIndex;
    }

    public LandAreaEntity landArea() {
        return landArea;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ClaimedChunkEntity that)) return false;
        if (id == null || that.id == null) return false;
        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return id == null ? System.identityHashCode(this) : Objects.hashCode(id);
    }
}

File: adapters/database/src/main/kotlin/net/onelitefeather/pandorascluster/database/models/ClaimedChunkEntity.kt
package net.onelitefeather.pandorascluster.database.models

import jakarta.persistence.*
import net.onelitefeather.pandorascluster.dbo.chunk.ClaimedChunkDBO

@Entity
@Table(name = "land_chunks")
data class ClaimedChunkEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = -1L,

    @Column
    val chunkIndex: Long = -1,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "land_id")
    val landEntity: LandEntity? = null
): ClaimedChunkDBO {

    @Override
    override fun toString(): String {
        return this::class.simpleName + "(id = $id , chunkIndex = $chunkIndex )"
    }

    override fun id(): Long? = id

    override fun chunkIndex(): Long = chunkIndex

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ClaimedChunkEntity) return false

        if (id != other.id) return false
        if (chunkIndex != other.chunkIndex) return false
        if (landEntity != other.landEntity) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id?.hashCode() ?: 0
        result = 31 * result + chunkIndex.hashCode()
        result = 31 * result + (landEntity?.hashCode() ?: 0)
        return result
    }
}